### DISCUSSION TITLE
Is there a way to measure performance or is there a benchmark for liboqs

### DISCUSSION THREAD

**Discussion by gabrofig** [open-quantum-safe/liboqs#1810] [Q&A] [answered]
Hi all,
As title says, is it possible to measure the performance in time of execution and memory usage (for example) of liboqs?

Thank you!

**Comment by SWilson4** [comment]
Hi @gabrofig, `liboqs` has a few test programs for measuring algorithm performance. See section 3 of the [quickstart guide](https://github.com/open-quantum-safe/liboqs?tab=readme-ov-file#linux-and-mac) for details.

**Reply by SWilson4** [reply]
Please note that I've converted this issue into a discussion as it doesn't raise an issue per se.

**Reply by gabrofig** [reply]
Thanks a lot, Wilson! I did manage to run the test refered in section 3. However, I only get information  that the test was successful like bellow:

test_mem.py ................................                                                            [100%]
============================================= 32 passed in 5.43s ==============================================

Does it mean that something might be missing in my build/setup? I would like to have information like memory usage or time of execution of each step of KEM, for example. I appreciate any feedback. Thanks!

**Reply by baentsch** [reply]
@gabrofig You may want to run the test in verbose mode or manually, e.g., as such `./test/speed_kem` in the build directory.

**Reply by gabrofig** [reply]
Thank you Michael. How could I run the test_kem_mem.c for example? I am trying to run directly vai terminal but it is not working.
Apologies for the noob question, but I must be doing something wrong.

**Reply by baentsch** [reply]
If memory serves, the program gives Usage help (if correctly built). Pretty similar to the other test programs (algname) and probably a test stage. Anyway, please always provide command tried, error message and what you expect.

**Reply by gabrofig** [reply]
I tried running test_mem.py directly in Visual Studio Code and got the following message

kem_name ML-KEM-1024
running subprocess
. > /home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem ML-KEM-1024 0
command: ['/home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem', 'ML-KEM-1024', '0']
result ------------------------------------- CompletedProcess(args=['/home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem', 'ML-KEM-1024', '0'], returncode=0, stdout=b'Configuration info\n==================\nTarget platform:  x86_64-Linux-6.5.0-35-generic\nCompiler:         gcc (11.4.0)\nCompile options:  [-Wa,--noexecstack;-O3;-fomit-frame-pointer;-fdata-sections;-ffunction-sections;-Wl,--gc-sections;-Wbad-function-cast]\nOQS version:      0.10.1-dev\nGit commit:       6ee5de25220e490cf7c895939c0fa001e5f8e2ce\nOpenSSL enabled:  Yes (OpenSSL 3.0.2 15 Mar 2022)\nAES:              NI\nSHA-2:            OpenSSL\nSHA-3:            C\nOQS build flags:  BUILD_SHARED_LIBS OQS_DIST_BUILD OQS_OPT_TARGET=generic CMAKE_BUILD_TYPE=Release \nCPU exts active:  AES AVX AVX2 BMI1 BMI2 PCLMULQDQ POPCNT SSE SSE2 SSE3\n================================================================================\nExecuting keygen for KEM ML-KEM-1024\n================================================================================\nStarting OQS_MEM_SECURE_FREE 1')
kem_name ML-KEM-1024
running subprocess
. > /home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem ML-KEM-1024 1
command: ['/home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem', 'ML-KEM-1024', '1']
result ------------------------------------- CompletedProcess(args=['/home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem', 'ML-KEM-1024', '1'], returncode=0, stdout=b'Configuration info\n==================\nTarget platform:  x86_64-Linux-6.5.0-35-generic\nCompiler:         gcc (11.4.0)\nCompile options:  [-Wa,--noexecstack;-O3;-fomit-frame-pointer;-fdata-sections;-ffunction-sections;-Wl,--gc-sections;-Wbad-function-cast]\nOQS version:      0.10.1-dev\nGit commit:       6ee5de25220e490cf7c895939c0fa001e5f8e2ce\nOpenSSL enabled:  Yes (OpenSSL 3.0.2 15 Mar 2022)\nAES:              NI\nSHA-2:            OpenSSL\nSHA-3:            C\nOQS build flags:  BUILD_SHARED_LIBS OQS_DIST_BUILD OQS_OPT_TARGET=generic CMAKE_BUILD_TYPE=Release \nCPU exts active:  AES AVX AVX2 BMI1 BMI2 PCLMULQDQ POPCNT SSE SSE2 SSE3\nttttt================================================================================\nExecuting encaps for KEM ML-KEM-1024\nttttt================================================================================\nStarting OQS_MEM_SECURE_FREE 1')
kem_name ML-KEM-1024
running subprocess
. > /home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem ML-KEM-1024 2
command: ['/home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem', 'ML-KEM-1024', '2']
result ------------------------------------- CompletedProcess(args=['/home/g339268/liboqs 2/liboqs/build/tests/test_kem_mem', 'ML-KEM-1024', '2'], returncode=0, stdout=b'Configuration info\n==================\nTarget platform:  x86_64-Linux-6.5.0-35-generic\nCompiler:         gcc (11.4.0)\nCompile options:  [-Wa,--noexecstack;-O3;-fomit-frame-pointer;-fdata-sections;-ffunction-sections;-Wl,--gc-sections;-Wbad-function-cast]\nOQS version:      0.10.1-dev\nGit commit:       6ee5de25220e490cf7c895939c0fa001e5f8e2ce\nOpenSSL enabled:  Yes (OpenSSL 3.0.2 15 Mar 2022)\nAES:              NI\nSHA-2:            OpenSSL\nSHA-3:            C\nOQS build flags:  BUILD_SHARED_LIBS OQS_DIST_BUILD OQS_OPT_TARGET=generic CMAKE_BUILD_TYPE=Release \nCPU exts active:  AES AVX AVX2 BMI1 BMI2 PCLMULQDQ POPCNT SSE SSE2 SSE3\n================================================================================\nExecuting decaps for KEM ML-KEM-1024\n================================================================================\nshared secrets are equal\n0000000000000000000test 1Starting OQS_MEM_SECURE_FREE 1')
======================================= test session starts =======================================
platform linux -- Python 3.10.12, pytest-6.2.5, py-1.10.0, pluggy-0.13.0
rootdir: /home/g339268/liboqs 2/liboqs/tests
plugins: xdist-2.5.0, forked-1.4.0
collected 32 items                                                                                

test_mem.py ................................                                                [100%]

======================================= 32 passed in 2.90s ========================================

From what I understood, this result confirms that KEM was executed successfully. However, I would like to see the results for the memory usage or time of execution of each steps (keygen, encap, decap) for example.

**Reply by baentsch** [reply]
> However, I would like to see the results for the memory usage or time of execution of each steps (keygen, encap, decap) for example.

The OQS test code is there to be used within suitable `valgrind` executions (tuned to what your particular interest is, stack, heap, whatever). I'm afraid there are no ready-made invocations for simple use (just for constant time code path testing). If you set up something that generates data of interest and value, please consider a PR @gabrofig.

**Reply by baentsch** [reply]
PS: Just remembered I did this many years ago: Take a look at https://github.com/open-quantum-safe/profiling/blob/main/perf/scripts/run_mem.py

**Comment by baentsch** [comment]
I'm sorry, I can't really help here as I don't have my computer around this week. Maybe @SWilson4 can help now. Otherwise I try to look into this next week.

⁣BlueMail for Android herunterladen ​

**Reply by gabrofig** [reply]
Thank you Michael, I will kindly ask Spencer in case he can support.
When I tried to build ARM64, with the docker buildx build -t oqs-perf:arm64 --platform=linux/arm64 . I got the following errors that I assume are due lack of permission to the docker (?)
![image](https://github.com/user-attachments/assets/b2b9fc57-78c9-4bc5-8834-e2fa2b0afafe)

Is this build necessary to use the run_mem.py script for KEM? If not, how can I use this script to assess performance for KEM?
I appreciate any feedback!

**Reply by gabrofig** [reply]
Just adding: I simply ran sudo docker run -v /home/tests:/opt/test/results oqs-perf and got some results as bellow before having the following error: Trying to mount S3...
./mount_s3.sh expects BUCKETSECRETS to authenticate to S3. Exiting without mount.
Couldn't mount S3 bucket. Not copying out test results.

![image](https://github.com/user-attachments/assets/af7585f2-901a-488e-aa71-f24edc1151bf)


I believe those results are the performance ones that I mentioned, correct?

**Comment by baentsch** [comment]
> I believe those results are the performance ones that I mentioned, correct?

I'm not sure I understand what/where you mentioned them (?)

> Couldn't mount S3 bucket. Not copying out test results.

That error is expected as you probably didn't pass the required access credentials for some S3 bucket where to store the results (and surely don't have the ones OQS uses :)

**Reply by gabrofig** [reply]
Thank you Michael. 
I mean that, for example KEM 1024, the command docker run -v /home/tests:/opt/test/results oqs-perf generated results for total time (s)m time (us): mean, pop stdev CPU cycles etc. I believe those are already at least some of the metrics for the KEM algorithm. I am only missing memory consumption in those tests. Is there a way to run only a test to measure the memory consumption?

Kind regards,

**Reply by baentsch** [reply]
> Is there a way to run only a test to measure the memory consumption?

Yes. Please dissect the [core test runtime script](https://github.com/open-quantum-safe/profiling/blob/main/perf/scripts/run-tests.sh).

**Comment by gabrofig** [comment]
I Think I found it. I ran locally ./run-tests.sh and after some path manipulation and printing partial results for test_kem_mem, I got the following results

![image](https://github.com/user-attachments/assets/b33053d9-6ff5-4ce9-be4e-8925f5d3a4f7)


Thanks a lot for all the support from you guys!

**Reply by baentsch** [reply, marked as answer]
UW. OK to mark any comment as answer and close the issue this way?

**Reply by gabrofig** [reply]
Done, thank you again!
